### PR TITLE
Allow true bidirectional streaming in tests

### LINKED ISSUE #1518: Integration test improvements
It would be nice to have end-to-end tests which were slightly less deterministic, to make it easier to have many streams and large bidirectional data flows going independently of each other.    This results in less deterministic tests but allows fuzzing corner cases for things like watermark limits and unexpected disconnects and resets.

Possibly a separate issue but likely predicated on this, it'd be nice to be able to do simpler things with less boilerplate - set a default backend response and client->sendSynchronousRequest, then inspect response headers or upstream headers after the fact. 

Finally, while I'm thinking about it, perhaps make an easy config for which test should apply to which languages so one doesn't have to edit multiple integration tests every time you want something tested for http/h2/h2-upstream.

... and as long as I'm dreaming, I'd like a pony.

### PR DESCRIPTION
**Pull request by alyssawilk**
I was halfway writing this as the fix to the macos flake when I realized I didn't want to fix it without debugging it first.  So that's *actually* fixed in #1888 but this would have made the flake go away (leaving the latent bug though) by decoupling client and backend events and letting upstream do its thing.

This fixes #1518  We detect more corner cases internally at Google by having upstream and downstream operate more or less independently, and this allows us to do the same for Envoy.

@htuch @dnoe for design critique of the autonomous stream.  This is different than our internal requests test framework (where we manually plumb all actions through the backend and connection to streams).  This version allows us to easily add more actions without plumbing them through 3 layers of hierarchy but we'd still need to do that work to trigger error behaviors (disconnects / timeouts / responses) before reading headers.

Signed-off-by: Alyssa Wilk <alyssar@chromium.org>

### CODE REVIEW

**Comment by alyssawilk** [comment]
+1 - I would have found it useful when I joined.
Want to gate this one on it or do it as a follow-up?  If follow-up is OK I'll file an issue to myself so it doesn't get lost.

File: test/integration/autonomous_upstream.cc [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  }
+}
+
+// Check all the special headers and send a customized response based on them.
+void AutonomousStream::sendResponse() {
+  Http::TestHeaderMapImpl headers(*headers_);
+
+  int32_t request_body_length = -1;
+  HeaderToInt(EXPECT_REQUEST_SIZE_BYTES, request_body_length, headers);
+  if (request_body_length > 0) {
```
Comment (htuch) [inline test/integration/autonomous_upstream.cc]: `>=`?

File: test/integration/fake_upstream.h [hunk 2]
```diff
@@ -291,6 +295,12 @@ class FakeUpstream : Logger::Loggable<Logger::Id::testing>, public Network::Filt
   bool createFilterChain(Network::Connection& connection) override;
   void set_allow_unexpected_disconnects(bool value) { allow_unexpected_disconnects_ = value; }
 
+protected:
+  std::mutex lock_;
```
Comment (htuch) [inline test/integration/fake_upstream.h]: Can you add a comment on what this mutex protects? Do you think there is value in using the TSAN annotations to validate this (given it's only test code)?

File: test/integration/autonomous_upstream.cc [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    sendResponse();
+  }
+}
+
+// Check all the special headers and send a customized response based on them.
+void AutonomousStream::sendResponse() {
+  Http::TestHeaderMapImpl headers(*headers_);
+
+  int32_t request_body_length = -1;
+  HeaderToInt(EXPECT_REQUEST_SIZE_BYTES, request_body_length, headers);
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply test/integration/autonomous_upstream.cc]: Yeah, that's what I mean by plumbing through - we'd have to add an API for each to AutonomousUpstream, AutonomousHttpConnection and AutonomousStream

The other nice benefit of doing this in headers is it ties the requests to responses cleanly.  Google style if we said "every 5th stream at the backend gets reset" it would be tricky associating the client streams to backend streams, since there's no consistent ordering guarantee.  Here because we do it in the stream headers, you can easily ensure the specific stream which requested reset will be reset.

File: test/integration/fake_upstream.h (hunk 2 above)
Comment (alyssawilk) [reply test/integration/fake_upstream.h]: Yeah, I was overenthusiastic.  Commented but not guarded until we switch to abseil mutex :-)

File: test/integration/autonomous_upstream.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply test/integration/autonomous_upstream.cc]: FWIW I like the header approach. I think it's really easy to reason about and easy to use.

**Comment by mattklein123** [comment]
> Want to gate this one on it or do it as a follow-up? If follow-up is OK I'll file an issue to myself so it doesn't get lost.

Sure, follow up for the docs is fine.

File: test/integration/autonomous_upstream.cc [hunk 4]
```diff
@@ -0,0 +1,77 @@
+#include "test/integration/autonomous_upstream.h"
+
+namespace Envoy {
+
+namespace {
+
+void HeaderToInt(const char header_name[], int32_t& return_int, Http::TestHeaderMapImpl& headers) {
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply test/integration/autonomous_upstream.cc]: I really want to get rid of the get response status thing in the prod code, but I'm still not sure about the best way to do it. I need to find time to look into it more.

File: test/integration/autonomous_upstream.cc [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  std::string header_value = headers.get_(header_name);
+  if (!header_value.empty()) {
+    return_int = atoi(header_value.c_str());
```
Comment (mattklein123) [inline test/integration/autonomous_upstream.cc]: Should we use StringUtil::atoul here though and verify/assert that the header is a valid int?
